File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

bool check_sum(long n);
int get_length(long n);
bool check_amex(int length, long cc);
bool check_visa(int length, long cc);
bool check_mastercard(int length, long cc);

int main(void)
{
    long cc_number = get_long("Number: ");
    if (!check_sum(cc_number))
    {
        printf("INVALID\n");
        return 0;
    }
    int length = get_length(cc_number);
    if (check_amex(length, cc_number))
    {
        printf("AMEX\n");
        return 0;
    }
    if (check_visa(length, cc_number))
    {
        printf("VISA\n");
        return 0;
    }
    if (check_mastercard(length, cc_number))
    {
        printf("MASTERCARD\n");
        return 0;
    }
    printf("INVALID\n");
    return 0;
}

int get_length(long n)
{
    int counter = 0;
    while (n > 0)
    {
        n /= 10;
        ++counter;
    }
    return counter;
}

bool check_sum(long n)
{
    int even = 0;
    int odd = 0;
    int check = 0;
    while (n > 0)
    {
        // odd numbers
        int x = n % 10;
        odd += x;
        n /= 10;

        // even numbers
        int y = (n % 10) * 2;
        while (y > 0)
        {
            even += y % 10;
            y /= 10;
        }
        n /= 10;
    }
    check = even + odd;
    if (!(check % 10 == 0))
    {
        return false;
    }
    return true;
}

bool check_amex(int length, long cc)
{
    if (!(length == 15))
    {
        return false;
    }
    int front_two = cc / 10000000000000;
    if ((front_two == 34 || front_two == 37))
    {
        return true;
    }
    return false;
}

bool check_visa(int length, long cc)
{
    if (!(length == 13 || length == 16))
    {
        return false;
    }
    if (length == 16 && cc / 1000000000000000 == 4)
    {
        return true;
    }
    if (length == 13 && (cc / 1000000000000 == 4))
    {
        return true;
    }
    return false;
}

bool check_mastercard(int length, long cc)
{
    if (!(length == 16))
    {
        return false;
    }
    int front_two = cc / 100000000000000;
    if ((front_two < 51 || front_two > 55))
    {
        return false;
    }
    return true;
}

File: pset6/sentimental-credit/credit.py
from cs50 import get_int


def get_creditcard_number() -> int:
    num = get_int('Number: ')
    return num


def checksum_check(x: int) -> bool:
    """Returns True if x passes luhn's algo."""
    assert isinstance(x, int), 'x should be an integer.'
    x = str(x)
    one, two = 0, 0
    while len(x) > 0:
        if len(x) == 1:
            one += int(x)
            break
        even, odd = int(x[-2]) * 2, int(x[-1])
        one += odd
        even = str(even)
        while len(even) > 0:
            two += int(even[-1])
            even = even[:-1]
        x = x[:-2]
    total_sum = one + two
    if total_sum % 10 == 0:
        return True
    else:
        return False


def check_visa(x: int) -> bool:
    # Checks the length
    if len(str(x)) != 13 and len(str(x)) != 16:
        return False
    first = int(str(x)[0])
    if first == 4:
        return True
    else:
        return False


def check_mastercard(x: int) -> bool:
    if len(str(x)) != 16:
        return False
    first_two = int(str(x)[:2])
    if first_two >= 51 and first_two < 56:
        return True
    return False


def check_amex(x: int) -> bool:
    if len(str(x)) != 15:
        return False
    first_two = int(str(x)[:2])
    if first_two != 34 and first_two != 37:
        return False
    return True


if __name__ == "__main__":
    cc = get_creditcard_number()
    if checksum_check(cc) == False:
        print("INVALID\n")
    else:
        if check_visa(cc) == True:
            print("VISA\n")
        elif check_mastercard(cc) == True:
            print("MASTERCARD\n")
        elif check_amex(cc) == True:
            print("AMEX\n")
        else:
            print("INVALID\n")
